File: android/src/main/kotlin/com/klarna/mobile/sdk/flutter/postpurchase/manager/KlarnaPostPurchaseSDKManager.kt
package com.klarna.mobile.sdk.flutter.postpurchase.manager

import com.klarna.mobile.sdk.flutter.PluginContext
import com.klarna.mobile.sdk.flutter.ResultError
import com.klarna.mobile.sdk.flutter.core.util.KlarnaParamMapper
import com.klarna.mobile.sdk.flutter.postpurchase.method.KlarnaPostPurchaseSDKMethod
import com.klarna.mobile.sdk.flutter.postpurchase.callback.KlarnaPostPurchaseSDKCallbackImpl
import com.klarna.mobile.sdk.api.postpurchase.KlarnaPostPurchaseSDK
import io.flutter.plugin.common.MethodChannel

internal class KlarnaPostPurchaseSDKManager(private val id: Int) {

    companion object {
        private const val NOT_INITIALIZED = "KlarnaPostPurchaseSDK is not initialized"

        fun notInitialized(result: MethodChannel.Result?) {
            result?.error(
                ResultError.KLARNA_POST_PURCHASE_SDK_ERROR.errorCode,
                NOT_INITIALIZED,
                "Call 'KlarnaPostPurchaseSDK.createInstance' before this."
            )
        }
    }

    private var postPurchaseSDK: KlarnaPostPurchaseSDK? = null
    private val postPurchaseSDKCallback: KlarnaPostPurchaseSDKCallbackImpl =
        KlarnaPostPurchaseSDKCallbackImpl(id)

    fun create(method: KlarnaPostPurchaseSDKMethod.Create, result: MethodChannel.Result?) {
        PluginContext.activity?.let { activity ->
            val environment = KlarnaParamMapper.getEnvironment(method.environment)
            val region = KlarnaParamMapper.getRegion(method.region)
            val resourceEndpoint = KlarnaParamMapper.getResourceEndpoint(method.resourceEndpoint)
            postPurchaseSDK = KlarnaPostPurchaseSDK(
                activity,
                environment,
                region,
                postPurchaseSDKCallback,
                resourceEndpoint
            )
            postPurchaseSDK?.returnURL = method.returnURL
        } ?: run {
            result?.error(
                ResultError.KLARNA_POST_PURCHASE_SDK_ERROR.errorCode,
                "Failed to construct.",
                null
            )
            return
        }
        result?.success(null)
    }

    fun initialize(method: KlarnaPostPurchaseSDKMethod.Initialize, result: MethodChannel.Result?) {
        postPurchaseSDK?.let {
            it.initialize(method.locale, method.purchaseCountry, method.design)
        } ?: run {
            notInitialized(result)
            return
        }
        result?.success(null)
    }

    fun authorizationRequest(
        method: KlarnaPostPurchaseSDKMethod.AuthorizationRequest,
        result: MethodChannel.Result?
    ) {
        postPurchaseSDK?.let {
            it.authorizationRequest(
                method.clientId,
                method.scope,
                method.redirectUri,
                method.locale,
                method.state,
                method.loginHint,
                method.responseType
            )
        } ?: run {
            notInitialized(result)
            return
        }
        result?.success(null)
    }

    fun renderOperation(
        method: KlarnaPostPurchaseSDKMethod.RenderOperation,
        result: MethodChannel.Result?
    ) {
        postPurchaseSDK?.let {
            it.renderOperation(method.operationToken, method.locale, method.redirectUri)
        } ?: run {
            notInitialized(result)
            return
        }
        result?.success(null)
    }

    fun destroy(method: KlarnaPostPurchaseSDKMethod.Destroy, result: MethodChannel.Result?) {
        postPurchaseSDK?.removeCallback(postPurchaseSDKCallback)
        postPurchaseSDK = null
        result?.success(null)
    }
}

File: packages/klarna_mobile_sdk_flutter/ios/klarna_mobile_sdk_flutter/Sources/klarna_mobile_sdk_flutter/PostPurchase/Manager/KlarnaPostPurchaseSDKManager.swift
import Flutter
import WebKit
import KlarnaMobileSDK

class KlarnaPostPurchaseSDKManager {
    
    static let NOT_INITIAILIZED = "KlarnaPostPurchaseSDK is not initialized"
    
    static func notInitialized(result: FlutterResult?) {
        result?(FlutterError.init(code: ResultError.klarnaPostPurchaseSDKError.rawValue, message: NOT_INITIAILIZED, details: "Call 'KlarnaPostPurchaseSDK.createInstance' before using this"))
    }
    
    let id: Int
    var postPurchaseSDK: KlarnaPostPurchaseSDK?
    var postPurchaseSDKEventListener: KlarnaPostPurchaseEventListener
    
    init(id: Int) {
        self.id = id
        self.postPurchaseSDKEventListener = KlarnaPostPurchaseSDKEventListenerImpl(id: id)
    }
    
    func create(method: KlarnaPostPurchaseSDKMethods.Create, result: @escaping FlutterResult) {
        let environment = KlarnaParamMapper.getEnvironmentOrDefault(param: method.environment)
        let region = KlarnaParamMapper.getRegionOrDefault(param: method.region)
        let resourceEndpoint = KlarnaParamMapper.getResourceEndpointOrDefault(param: method.resourceEndpoint)
        postPurchaseSDK = KlarnaPostPurchaseSDK(environment: environment, region: region, resourceEndpoint: resourceEndpoint, listener: postPurchaseSDKEventListener)
        if let returnURLString = method.returnURL {
            postPurchaseSDK?.returnURL = URL(string: returnURLString)
        }
        result(nil)
    }
    
    func initialize(method: KlarnaPostPurchaseSDKMethods.Initialize, result: @escaping FlutterResult) {
        if let postPurchaseSDK = postPurchaseSDK {
            postPurchaseSDK.initialize(locale: method.locale, purchaseCountry: method.purchaseCountry, design: method.design)
            result(nil)
        } else {
            KlarnaPostPurchaseSDKManager.notInitialized(result: result)
        }
    }
    
    func authorizationRequest(method: KlarnaPostPurchaseSDKMethods.AuthorizationRequest, result: @escaping FlutterResult) {
        if let postPurchaseSDK = postPurchaseSDK {
            postPurchaseSDK.authorizationRequest(clientId: method.clientId, scope: method.scope, redirectUri: method.redirectUri, locale: method.locale, state: method.state, loginHint: method.loginHint, responseType: method.responseType)
            result(nil)
        } else {
            KlarnaPostPurchaseSDKManager.notInitialized(result: result)
        }
    }
    
    func renderOperation(method: KlarnaPostPurchaseSDKMethods.RenderOperation, result: @escaping FlutterResult) {
        if let postPurchaseSDK = postPurchaseSDK {
            postPurchaseSDK.renderOperation(operationToken: method.operationToken, locale: method.locale, redirectUri: method.redirectUri)
            result(nil)
        } else {
            KlarnaPostPurchaseSDKManager.notInitialized(result: result)
        }
    }
    
    func destroy(method: KlarnaPostPurchaseSDKMethods.Destroy, result: @escaping FlutterResult) {
        postPurchaseSDK = nil
        result(nil)
    }
    
}
